src/systems.rs

```diff
@@ -321,25 +321,43 @@ pub(crate) fn particle_transform(
     mut particle_query: Query<(
         &Particle,
         &Lifetime,
-        &Direction,
+        &mut Velocity,
         &mut DistanceTraveled,
-        &mut Speed,
         &mut Transform,
     )>,
     time: Res<Time>,
 ) {
     particle_query.par_iter_mut().for_each_mut(
-        |(particle, lifetime, direction, mut distance, mut speed, mut transform)| {
+        |(particle, lifetime, mut velocity, mut distance, mut transform)| {
             let lifetime_pct = lifetime.0 / particle.max_lifetime;
-            if particle.use_scaled_time {
-                speed.0 +=
-                    particle.acceleration.at_lifetime_pct(lifetime_pct) * time.delta_seconds();
-                transform.translation += direction.0 * speed.0 * time.delta_seconds();
-            } else {
-                speed.0 +=
-                    particle.acceleration.at_lifetime_pct(lifetime_pct) * time.raw_delta_seconds();
-                transform.translation += direction.0 * speed.0 * time.raw_delta_seconds();
+
+            let delta_time = match particle.use_scaled_time {
+                true => time.delta_seconds(),
+                false => time.raw_delta_seconds(),
+            };
+
+            // The velocity_direction variable is needed due to its relation with the acceleration being a f32 and not a Vec3
+            let velocity_direction = velocity.0.normalize();
+
+            // Apply acceleration
+            velocity.0 +=
+                velocity_direction
+                * particle.acceleration.at_lifetime_pct(lifetime_pct)
+                * delta_time;
+            
+            // Apply drag
+            let current_drag = particle.drag.at_lifetime_pct(lifetime_pct);
+            if current_drag > 0.0 {
+                let drag_force =
+                    ( velocity.0.length() * velocity.0.length() )
```

Comment:
We should use `.length_squared()` here. Computing `.length()` requires doing a very expensive square root as the final step. So `.length_squared()` isn't  `length * length`, it's the length formula without the square root step, which makes it substantially faster.